pmndrs/zustand
# Rewrite `TakeTwo`, `SkipTwo` and `Mutate` to make them work for older ts versions too

Fixes #1013. Superceeds #1275.

## What is the issue?

The core of the issue is this: TypeScript changed subtyping rules of tuples with optional elements (probably) in v4.4.4. The type `Test` in the following snippet resolves to `"yes"` in v4.3.5 and before but it resolves to `"no"` in 4.4.4 and after.

```ts
type Test = [string, number?] extends [unknown] ? "yes" : "no"
```
This should most probably be intentional as the change is so old and 4.4.4 also happens to be the version that came with `--exactOptionalProperties` so it's possible it touched upon tuples with optional element. It's also possible that this is a "fix" or that it increases the "strictness" of the subtyping rules. (Also turning `--exactOptionalProperties` off or on does not have any effect on the subtyping.)

Now how does it affect us? We have a type `TakeTwo<T>` that does a type-level equivalent of...

```js
const takeTwo = _x => {
  let y = [], x = [..._x];
  while (y.length < 2) y.push(x.shift())
  return y
}

// This is similar to `.slice(0, 2)` except we fill the remaining spaces with `undefined`s
// so that the final result is of length 2.
```

Now the problem is `TakeTwo` doesn't work for pre-4.4.4 version...

```ts
type TakeTwo<T> =
  T extends [] ? [undefined, undefined] :
  T extends [unknown] ? [...a0: T, a1: undefined] :
  T extends [unknown?] ? [...a0: T, a1: undefined] :
  T extends [unknown, unknown] ? T :
  T extends [unknown, unknown?] ? T :
  T extends [unknown?, unknown?] ? T :
  T extends [infer A0, infer A1, ...unknown[]] ? [A0, A1] :
  T extends [infer A0, (infer A1)?, ...unknown[]] ? [A0, A1?] :
  T extends [(infer A0)?, (infer A1)?, ...unknown[]] ? [A0?, A1?] :
  never

type Test = TakeTwo<[state: number, replace?: boolean]>
// Test is [state: number, replace: boolean | undefined, a1: undefined] <4.4.4
// and [state: number, replace: boolean | undefined] >= 4.4.4
```

So now even though our code is "correct" it doesn't work with pre-4.4.4. Same is the case for `SkipTwo` that does a type-level `.slice(2)`.

## How do we solve it?

Now in some sense we can't write a code that works for both versions because typescript changed the math itself.

Imagine you have a function `f` as `const f = (a, b) => a <> b`. Now in wild-land-4.3 `<>` did subtraction, but now in wild-lang-4.4 `<>` does addition. Now you can't write a function `f` that does addition in both wild-lang-4.3 and wild-land-4.4.

So if we want to write `TakeTwo` that works for both pre-change and post-change versions, we'll have to use only those rules that are same in both, some subtyping rules for tuples with optional element are not same so we can't use them.

What is consistent in both is the length property, eg `[string, number?]["length"]` is `1 | 2` for both versions. So we can compare the lengths instead of comparing the tuples themselves...

```ts
type Test = [string, number?] extends { length: 1 | 2 }  ? "yes" : "no"
// Test is "yes" in <4.4.4 and >=4.4.4
```

So the new `TakeTwo` looks like this...

```ts
type TakeTwo<T> =
  T extends { length: 0 } ? [undefined, undefined] :
  T extends { length: 1 } ? [...a0: Cast<T, unknown[]>, a1: undefined] :
  T extends { length: 0 | 1 } ? [...a0: Cast<T, unknown[]>, a1: undefined] :
  T extends { length: 2 } ? T :
  T extends { length: 1 | 2 } ? T :
  T extends { length: 0 | 1 | 2 } ? T :
  T extends [infer A0, infer A1, ...unknown[]] ? [A0, A1] :
  T extends [infer A0, (infer A1)?, ...unknown[]] ? [A0, A1?] :
  T extends [(infer A0)?, (infer A1)?, ...unknown[]] ? [A0?, A1?] :
  never

type Cast<T, U> =
  T extends U ? T : U

type Test = TakeTwo<[state: number, replace?: boolean]>
// Test is [state: number, replace?: boolean] in both <4.4.4 and >= 4.4.4
```

(The last three cases remain the same as the subtyping rule is consistent for them, and nor there's another way of writing them.)

And we make a similar change for `SkipTwo` also.

## How do we test it

Now usually libraries only target the latest typescript version, or whatever version the repo is running on, but here we have to target an old version.

This usualy would have been simple and we could test it with a script like so...

```sh
tsc --noEmit # type-check with current ts version
./node_modules/typescript-4.3.5/bin/tsc --noEmit # type-check with old ts version
```

But now the problem is the version 4.3.5 is so old that there would be some current tests that would be failing because we're using new features, eg `--exactOptionalProperties`. So now those type-check errors are expected, so we can't just rely on the exit code. What we'll have to do see if the tests type-check except some errors that we expect.

To do this I'm using inline snapshots. That is the `oldTsc --noEmit` output should only contain type errors we expect.

To confirm that this change in `TakeTwo` and `SkipTwo` fixes the problem, see the snapshot in first commit, it has 38 errors, but now after we make this change, the new snapshot in the second commit only has 6 expected errors that we can't fix because it's an older tsc that doesn't support the new features and changes.
